app/controllers/composer_controller.rb

```diff
@@ -0,0 +1,165 @@
+# frozen_string_literal: true
+
+class ComposerController < ApplicationController
+  requires_login
+
+  def mentions
+    names = params.require(:names)
+    raise Discourse::InvalidParameters.new(:names) if !names.kind_of?(Array) || names.size > 20
+
+    if params[:topic_id].present?
+      topic = Topic.find_by(id: params[:topic_id])
+      guardian.ensure_can_see!(topic)
+    end
+
+    # allowed_names is necessary just for new private messages.
+    allowed_names = if params[:allowed_names].present?
+      params[:allowed_names].split(',') << current_user.username
+    else
+      []
+    end
+
+    user_reasons = {}
+    group_reasons = {}
+    here_count = nil
+
+    users = User
+      .where(username_lower: names.map(&:downcase), staged: false)
+      .index_by(&:username_lower)
```

Comment:
I _think_ you can use `not_staged` so that you don't have two conditions in the `where` which might make it easy to overlook

```suggestion
    users = User
      .not_staged
      .where(username_lower: names.map(&:downcase))
      .index_by(&:username_lower)
```